src/Fuzz.elm

```diff
@@ -714,84 +543,25 @@ andMap =
-                Gen genVal ->
-                    Gen <| Random.andThen (transform >> Internal.unpackGenVal) genVal
-
-                Shrink genTree ->
-                    Shrink <| andThenRoseTrees transform genTree
-
-                InvalidFuzzer reason ->
-                    InvalidFuzzer reason
-        )
-
-
-andThenRoseTrees : (a -> Fuzzer b) -> Generator (RoseTree a) -> Generator (RoseTree b)
-andThenRoseTrees transform genTree =
-    genTree
-        |> Random.andThen
-            (\(Rose root branches) ->
-                let
-                    genOtherChildren : Generator (LazyList (RoseTree b))
-                    genOtherChildren =
-                        branches
-                            |> Lazy.List.map (\rt -> RoseTree.map (transform >> Internal.unpackGenTree) rt |> unwindRoseTree)
-                            |> unwindLazyList
-                            |> Random.map (Lazy.List.map RoseTree.flatten)
-                in
-                Random.map2
-                    (\(Rose trueRoot rootsChildren) otherChildren ->
-                        Rose trueRoot (Lazy.List.append rootsChildren otherChildren)
-                    )
-                    (Internal.unpackGenTree (transform root))
-                    genOtherChildren
-            )
-
-
-unwindRoseTree : RoseTree (Generator a) -> Generator (RoseTree a)
-unwindRoseTree (Rose genRoot lazyListOfRoseTreesOfGenerators) =
-    case Lazy.List.headAndTail lazyListOfRoseTreesOfGenerators of
-        Nothing ->
-            Random.map RoseTree.singleton genRoot
-
-        Just ( Rose gen children, moreList ) ->
-            Random.map4 (\a b c d -> Rose a (Lazy.List.cons (Rose b c) d))
-                genRoot
-                gen
-                (Lazy.List.map unwindRoseTree children |> unwindLazyList)
-                (Lazy.List.map unwindRoseTree moreList |> unwindLazyList)
-
-
-unwindLazyList : LazyList (Generator a) -> Generator (LazyList a)
-unwindLazyList lazyListOfGenerators =
-    case Lazy.List.headAndTail lazyListOfGenerators of
-        Nothing ->
-            Random.constant Lazy.List.empty
-
-        Just ( head, tail ) ->
-            Random.map2 Lazy.List.cons head (unwindLazyList tail)
+andThen =
+    Internal.andThen
 
 
 {-| Conditionally filter a fuzzer to remove occasional undesirable
 input. Takes a limit for how many retries to attempt, and a fallback
 function to, if no acceptable input can be found, create one from an
 unacceptable one. Also takes a condition to determine if the input is
 acceptable or not, and finally the fuzzer itself.
-
 A good number of max retires is ten. A large number of retries might
 blow the stack.
-
 -}
 conditional : { retries : Int, fallback : a -> a, condition : a -> Bool } -> Fuzzer a -> Fuzzer a
 conditional { retries, fallback, condition } fuzzer =
     if retries <= 0 then
         map fallback fuzzer
     else
         fuzzer
-            |> andThen
+            |> Internal.andThenNoHistory
```

Comment:
I'm not a hundred percent sure this is valid, but it is a huge performance improvement if so. Regular `andThen` squares the amount of shrunken values for each level of nesting here, which over a maximum of 10 iterations can be punative. This version uses only the shrunken values from the fuzzer returned from the bind function `a -> Fuzzer b` and ignores the history.